File: buildframework/helium/sf/java/ccmtask/src/com/nokia/helium/ccmtask/ant/taskdefs/CcmTask.java
/*
* Copyright (c) 2007-2008 Nokia Corporation and/or its subsidiary(-ies).
* All rights reserved.
* This component and the accompanying materials are made available
* under the terms of the License "Eclipse Public License v1.0"
* which accompanies this distribution, and is available
* at the URL "http://www.eclipse.org/legal/epl-v10.html".
*
* Initial Contributors:
* Nokia Corporation - initial contribution.
*
* Contributors:
*
* Description: 
*
*/
 
package com.nokia.helium.ccmtask.ant.taskdefs;

import java.io.File;
import java.io.IOException;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.ArrayList;
import java.util.List;
import java.util.Vector;
import java.util.jar.JarFile;
import java.util.zip.ZipEntry;

import org.python.util.PythonInterpreter;

import com.nokia.helium.ccmtask.ant.commands.AddTask;
import com.nokia.helium.ccmtask.ant.commands.CcmCommand;
import com.nokia.helium.ccmtask.ant.commands.ChangeReleaseTag;
import com.nokia.helium.ccmtask.ant.commands.Checkout;
import com.nokia.helium.ccmtask.ant.commands.Close;
import com.nokia.helium.ccmtask.ant.commands.Reconcile;
import com.nokia.helium.ccmtask.ant.commands.Role;
import com.nokia.helium.ccmtask.ant.commands.Snapshot;
import com.nokia.helium.ccmtask.ant.commands.Synchronize;
import com.nokia.helium.ccmtask.ant.commands.Update;
import com.nokia.helium.ccmtask.ant.commands.Workarea;
import com.nokia.helium.ccmtask.ant.types.SessionSet;
import com.nokia.helium.ccmtask.ant.commands.CreateReleaseTag;
import com.nokia.helium.ccmtask.ant.commands.DeleteReleaseTag;
import org.apache.tools.ant.BuildException;
import org.apache.tools.ant.Task;

/**
 * Synergy task.
 * <pre>
 * &lt;hlm:createSessionMacro database="to1tobet" reference="test.session" /&gt;
 * &lt;hlm:ccm verbose="false"&gt;
 *     &lt;hlm:sessionset refid="test.session" /&gt;
 *     &lt;hlm:addtask folder="tr1test1#2079"&gt;
 *         &lt;task name="tr1test1#5310" /&gt;
 *     &lt;/hlm:addtask&gt;
 *     &lt;hlm:snapshot project="helium-to1tobet#helium_3.0:project:vc1s60p1#1" dir="c:\test" fast="true" recursive="true" /&gt;
 *     &lt;hlm:synchronize project="helium-to1tobet#helium_3.0:project:vc1s60p1#1" recursive="true" /&gt;
 *     &lt;hlm:close /&gt;
 * &lt;/hlm:ccm&gt;
 * </pre>
 * @ant.task category="SCM"
 */
public class CcmTask extends Task
{
    private String username;

    private String password;

    private List<CcmCommand> commands = new ArrayList<CcmCommand>();
    private Vector<SessionSet> sessionSets = new Vector<SessionSet>();
    private boolean verbose;

    public void setVerbose(boolean value) {
        verbose = value;
    }
    
    public boolean getVerbose() {
        return verbose;
    }

    public String getUsername()
    {
        return username;
    }

    public final void setUsername(final String username)
    {
        this.username = username;
    }

    public String getPassword()
    {
        return password;
    }

    public final void setPassword(final String password)
    {
        this.password = password;
    }

    public CcmCommand[] getCommands()
    {
        return commands.toArray(new CcmCommand[commands.size()]);
    }

    public void addUpdate(Update a)
    {
        addCommand(a);
    }
    
    public void addSynchronize(Synchronize a)
    {
        addCommand(a);
    }
    
    public void addReconcile(Reconcile a)
    {
        addCommand(a);
    }
    
    public void addSnapshot(Snapshot a)
    {
        addCommand(a);
    }
    
    public void addChangeReleaseTag(ChangeReleaseTag a)
    {
        addCommand(a);
    }

    public void addCreateReleaseTag(CreateReleaseTag a)
    {
        addCommand(a);
    }

    public void addDeleteReleaseTag(DeleteReleaseTag a)
    {
        addCommand(a);
    }

    public void addCheckout(Checkout a)
    {
        addCommand(a);
    }

    public void addWorkarea(Workarea a)
    {
        addCommand(a);
    }

    public void addRole(Role a)
    {
        addCommand(a);
    }

    private void addCommand(CcmCommand cmd)
    {
        cmd.setTask(this);
        commands.add(cmd);
    }

    public AddTask createAddTask()
    {
        AddTask cmd = new AddTask();
        addCommand(cmd);
        return cmd;
    }

    public Role createRole()
    {
        Role cmd = new Role();
        addCommand(cmd);
        return cmd;
    }

    public Close createClose()
    {
        Close cmd = new Close();
        addCommand(cmd);
        return cmd;
    }

    public SessionSet createSessionSet()
    {
        SessionSet sessionSet = new SessionSet();
        sessionSets.add(sessionSet);
        return sessionSet;
    }
    
    public SessionSet[] getSessionSets() {
        SessionSet[] result = new SessionSet[sessionSets.size()];
        sessionSets.copyInto(result);
        return result; 
    }

    public final void execute()
    {
        File jar = getJarFile();
        if (jar == null) {
            throw new BuildException("Could not find the jar file for class " + this.getClass().getCanonicalName());
        }
        try {
            JarFile jarFile = new JarFile(jar);
            String entryName = this.getClass().getPackage().getName().replace('.', '/') + "/ccmtask.py";
            ZipEntry entry = jarFile.getEntry(entryName);
            if (entry == null) {
                throw new BuildException("CcmTask internal error: Could not find the following entry: " + entryName);
            }
            PythonInterpreter pi = new PythonInterpreter();
            pi.set("java_ccmtask", this);
            pi.set("project", getProject());
            pi.execfile(jarFile.getInputStream(entry), "ccmtask.py");
        } catch (IOException e) {
            throw new BuildException(e.getMessage(), e);
        }
    }
    
    /**
     * Returns the jar file name containing this class
     * @return a File object or null if not found.
     * @throws IMakerException
     */
    protected File getJarFile() {
        URL url = this.getClass().getClassLoader().getResource(this.getClass().getName().replace('.', '/') + ".class");
        if (url.getProtocol().equals("jar") && url.getPath().contains("!/")) {
            String fileUrl = url.getPath().split("!/")[0];
            try {
                return new File(new URL(fileUrl).getPath());
            } catch (MalformedURLException e) {
                throw new BuildException("Error determining the jar file where "
                        + this.getClass().getName() + " is located.", e);
            }
        }
        return null;
    }
}

File: buildframework/helium/sf/java/ccmtask/src/com/nokia/helium/ccmtask/ant/taskdefs/ccmtask.py
#============================================================================ 
#Name        : ccmtask.py 
#Part of     : Helium 

#Copyright (c) 2009 Nokia Corporation and/or its subsidiary(-ies).
#All rights reserved.
#This component and the accompanying materials are made available
#under the terms of the License "Eclipse Public License v1.0"
#which accompanies this distribution, and is available
#at the URL "http://www.eclipse.org/legal/epl-v10.html".
#
#Initial Contributors:
#Nokia Corporation - initial contribution.
#
#Contributors:
#
#Description:
#===============================================================================
""" Script to process Ant <ccm> task commands. """

import ccm
import ccm.extra
import sys
import logging
import ant
import string

logging.basicConfig(level=logging.INFO)
antlogger = ant.AntHandler(java_ccmtask)
rlogger = logging.getLogger('')
# replacing default handler by Ant one.
if len(rlogger.handlers) > 0:
    rlogger.removeHandler(rlogger.handlers[0])
rlogger.addHandler(antlogger)
logging.getLogger("ccm").setLevel(logging.INFO)
logger = logging.getLogger("ccm.ant")
session = None

def execute_update(session, command):
    java_ccmtask.log(str('Updating project: ' + command.getProject()))
    project = session.create(command.getProject())
    project.update()

def execute_role(session, command):
    if not command.getRole():
        raise Exception("The 'role' attribute has not been defined.")
    java_ccmtask.log(str('Updating role: ' + command.getRole()))
    session.role = command.getRole()
    
def execute_synchronize(session, command):
    if not command.getProject():
        raise Exception("The 'project' attribute has not been defined.")
    java_ccmtask.log(str('Synchronizing project: ' + command.getProject()))
    java_ccmtask.log(str("Recursive: %s" % command.getRecursive()))
    project = session.create(command.getProject())
    project.sync(command.getRecursive())    
        
def execute_reconcile(session, command):
    if not command.getProject():
        raise Exception("The 'project' attribute has not been defined.")
    java_ccmtask.log(str('Reconciling project: ' + command.getProject()))
    project = session.create(command.getProject())
    project.reconcile()

def execute_snapshot(session, command):
    if not command.getProject():
        raise Exception("The 'project' attribute has not been defined.")
    if not command.getDir():
        raise Exception("The 'dir' attribute has not been defined.")
    java_ccmtask.log(str('Snapshot of project: ' + command.getProject() + ' to ' + command.getDir()))
    project = session.create(command.getProject())
    if command.getFast()== True:
        ccm.extra.FastSnapshot(project, command.getDir())        
    else:
        project.snapshot(command.getDir(), command.getRecursive())

def execute_changereleasetag(session, command):
    if not command.getFolder():
        raise Exception("The 'folder' attribute has not been defined.")
    if not command.getReleaseTag():
        raise Exception("The 'releaseTag' attribute has not been defined.")
    java_ccmtask.log(str('Changing release tag for all tasks in the folder : ' + command.getFolder() + ' as ' + command.getReleaseTag()))
    #Search all task from the folder    
    folder = session.create("Folder " + str(command.getFolder()))
    for task in folder.tasks:
        if task.release != str(command.getReleaseTag()):
            task.release = str(command.getReleaseTag())
            
def execute_checkout(session, command):
    if not command.getProject():
        raise Exception("The 'project' attribute has not been defined.")
    java_ccmtask.log(str('Checking out project: ' + command.getProject() + ' with release tag ' + command.getRelease()))
    project = session.create(command.getProject())
    if (command.getRelease()):
        if command.getWa():
            wa = command.getWa()
        else:
            wa = None
        if command.getRecursive():
            recursive = command.getRecursive()
        else:
            recursive = None
        if command.getRelative():
            relative = command.getRelative()
        else:
            relative = None
        if command.getVersion():
            version = command.getVersion()
        else:
            version = None
        if command.getPurpose():
            purpose = command.getPurpose()
        else:
            purpose = None
        project.checkout(session.create(command.getRelease()), version, purpose, recursive)

def execute_createreleasetag(session, command):
    java_ccmtask.log(str('creating a release tag'))
    if not command.getProject():
        raise Exception("The 'project' attribute has not been defined.")
    project = session.create(command.getProject())
    if (command.getNewTag()):
        new_release_tag = command.getNewTag();
        project.create_release_tag(command.getRelease(), new_release_tag)

def execute_deletereleasetag(session, command):
    java_ccmtask.log(str('deleting a release tag'))
    if not command.getProject():
        raise Exception("The 'project' attribute has not been defined.")
    project = session.create(command.getProject())
    if (command.getNewTag()):
        new_release_tag = command.getNewTag();
        project.delete_release_tag(command.getRelease(), new_release_tag)

def execute_workarea(session, command):
    if not command.getProject():
        raise Exception("The 'project' attribute has not been defined.")
    java_ccmtask.log(str('Modifying work area for the project : ' + command.getProject()))
    project = session.create(command.getProject())
    if command.getMaintain():
        maintain = command.getMaintain()
    else:
        maintain = None
    if command.getRecursive():
        recursive = command.getRecursive()
    else:
        recursive = None
    if command.getRelative():
        relative = command.getRelative()
    else:
        relative = None
    if command.getPath():
        path = command.getPath()
    else:
        path = None
    if command.getPst():
        pst = command.getPst()
    else:
        pst = None
    if command.getWat():
        wat = command.getWat()
    else:
        wat = None
    project.work_area(maintain, recursive, relative, path, pst, wat)

def execute_addtask(session, command):
    if command.getFolder() != None:
        tasks = []        
        ccmfolder = session.create("Folder " + command.getFolder())
        role = find_folder_information(session, command)
        if role == "build_mgr":
            java_ccmtask.log("Changing role to %s" % role)
            session.role = "build_mgr"

        folder = session.create("Folder %s" % str(command.getFolder()))
        for task in folder.tasks:
            folder.remove(task)
        
        java_ccmtask.log("Adding task to folder %s" % ccmfolder)
        tasks.extend(map(lambda task: session.create("Task " + task.getName()), command.getTasks()))
        for task in tasks:
            ccmfolder.append(task)
    
def find_folder_information(session, command):
    cmdline = "folder -sh i " + command.getFolder()    
    result = session.execute(cmdline)
    result_array = str(result).split('\r')
    for ldata in result_array :        
        if ldata.find('Writable By:') > 0:
            wb,wd = ldata.split(':',1)
            if wd.lstrip() == "Build Manager":
                return "build_mgr"
            else:
                return "developer"

def execute_exists(session, command):
    fpn = None
    if command.getObject() != None:
        fpn = command.getObject()
    elif command.getFolder() != None:
		fpn = "Folder " + command.getFolder()
    elif command.getTask() != None:
		fpn = "Task " + command.getTask()

    if fpn == None:
        raise Exception("You need to define either a 'task'/'folder'/'object' attribute.")
    ccmo = session.create(fpn)
    if ccmo.exists():
        java_ccmtask.log("'%s' exists." % ccmo)
    else:
        raise Exception("Could not find '%s'." % ccmo)

def execute_close(session, command):
    java_ccmtask.log(str("Closing session %s." % str(session)))
    if hasattr(session, 'close_on_exit'):
        session.close_on_exit = True    
    session.close()
    
def referenceToObject(obj):
    if obj.isReference() == 1:
        ref = project.getReference(str(obj.getRefid().getRefId()))
        if ref == None:
            raise Exception("Could not find reference '%s'" % str(obj.getRefid().getRefId()))
        return ref
    else:
        return obj        
        
sessionIds = [] 
for session_set in java_ccmtask.getSessionSets():
    sessionIds.extend(map(lambda session: session.getAddr(), referenceToObject(session_set).getSessions()))

print "Session list: ", sessionIds
if len(sessionIds) > 0:
    sessionid = sessionIds.pop()
    session = ccm.Session(username=None, engine=None, dbpath=None, ccm_addr=sessionid, close_on_exit=False)
else:
    username = java_ccmtask.getUsername()
    password = java_ccmtask.getPassword()
    session = ccm.open_session(username=username, password=password)

if java_ccmtask.getVerbose() == 1:
    logging.getLogger("ccm").setLevel(logging.DEBUG)

#print dir(sys.modules['__main__'] )
ccm_commands = java_ccmtask.getCommands()
for command in ccm_commands:
    print "Running command '%s'" % command.getName()
    method_name = 'execute_' + command.getName()
    method = sys.modules['__main__'].__dict__[method_name]
    try:
        method(session, command)
    except Exception, e:
        if java_ccmtask.getVerbose() == 1:
            import traceback
            logger.error(traceback.print_exc(file=sys.stdout))
            if 'result' in e:
                logger.error(e.result)
        logger.error(e)
        raise e
